### PR TITLE
Adds scoverage support

### LINKED ISSUE #411: Set up code coverage
As per https://github.com/non/cats/commit/fa5182cff105157de0d16920f314916f24cd70db.

### PR DESCRIPTION
**Pull request by khronnuz**
Fixes #411 

Preview of how the report looks like is at https://codecov.io/github/khronnuz/shapeless?ref=9a4bc2936bff479547e4cd2bcfbf52c9a78ad163&view=all

### CODE REVIEW

**Comment by milessabin** [comment]
Thanks!

It looks like this will have to be conditional on the Scala version. I think it makes sense to only run the report for the Scala version on the main line of development, ie. 2.11.x currently. 2.12.x isn't fully supported yet, and 2.10.x has had to have some tests disabled because they only became supported in 2.11.x.

But otherwise LGTM :-)

**Comment by milessabin** [comment]
Does an scoverage account need to be created for milessabin/shapeless?

**Comment by milessabin** [comment]
Oh, I also noticed that there are a few test files being included in the coverage report (ie. the jvm 2.11+ specific ones).

**Comment by khronnuz** [comment]
@milessabin I will address the issues you mentioned re: test included in reports and only testing scala 2.11. Yes, you need to go to codecov.io to create an account and authorise the app, but it is a very simple process. Can be done in 2 minutes.

**Comment by khronnuz** [comment]
I have created https://github.com/scoverage/sbt-scoverage/issues/160 for the fact that tests are being included. I will try to exclude them using exclusion patterns until this is fixed.

**Comment by khronnuz** [comment]
Hmm, I have updated the script to only run coverage on scala 2.11.x, but it still tries to load plugins. I will look at this.

**Comment by milessabin** [comment]
I clicked and authorized and now there's this: https://codecov.io/github/milessabin/shapeless.

**Comment by milessabin** [comment]
Are we still blocked waiting on Scoverage to support 2.12.x?

**Comment by khronnuz** [comment]
Hey Miles. Still no movement on the sbt-scoverage project (no new release or reply to the ticket created). So, unfortunately, yes.

**Comment by milessabin** [comment]
Did you try @dwijnand hacky workaround to exclude it on the 2.12.x build?

**Comment by InTheNow** [comment]
@khronnuz I may have found a solution to this - the PR will need rebasing on master first.

```diff
--- build.sbt   2016-05-11 13:38:11.654195896 +0200
***************
*** 72,77 ****
--- 72,93 ----
    )
  }
  
+ val cmdlineProfile = sys.props.getOrElse("sbt.profile", default = "")
+ 
+ def profile(crossProject: CrossProject) = cmdlineProfile match {
+   case "2.12.x" =>
+     crossProject
+     .jsConfigure(_.disablePlugins(scoverage.ScoverageSbtPlugin))
+     .jvmConfigure(_.disablePlugins(scoverage.ScoverageSbtPlugin))
+ 
+   case _ => crossProject
+ }
+ 
+ def profile: Project ⇒ Project = p => cmdlineProfile match {
+   case "2.12.x" => p.disablePlugins(scoverage.ScoverageSbtPlugin)
+   case _ => p
+ }
+ 
  lazy val commonJsSettings = Seq(
    scalacOptions += {
      val tagOrHash =
***************
*** 95,100 ****
--- 111,117 ----
    releaseSettings ++ scoverageSettings
  
  lazy val root = project.in(file("."))
+   .configure(profile)
    .aggregate(coreJS, coreJVM)
    .dependsOn(coreJS, coreJVM)
    .settings(coreSettings:_*)
***************
*** 110,115 ****
--- 127,133 ----
  
  lazy val core = crossProject.crossType(CrossTypeMixed)
    .configure(configureJUnit)
+   .configure(profile)
    .settings(moduleName := "shapeless")
    .settings(coreSettings:_*)
    .configure(buildInfoSetup)
***************
*** 126,131 ****
--- 144,150 ----
  
  lazy val scratch = crossProject.crossType(CrossType.Pure)
    .configure(configureJUnit)
+   .configure(profile)
    .dependsOn(core)
    .settings(moduleName := "scratch")
    .settings(coreSettings:_*)
***************
*** 146,151 ****
--- 165,171 ----
  
  lazy val examples = crossProject.crossType(CrossType.Pure)
    .configure(configureJUnit)
+   .configure(profile)
    .dependsOn(core)
    .settings(moduleName := "examples")
    .settings(runAllIn(Compile))
```

and then in `.travis.yml`:
```bash
if [[ "$TRAVIS_SCALA_VERSION" == 2.12.* ]]; then
    ${SBT} -Dsbt.profile="2.12.x" validate ${PUBLISH}
else
    eval $COVERAGE && ${SBT} clean validate ${PUBLISH}
fi
```

**Comment by khronnuz** [comment]
@InTheNow sorry, I missed the notification for your message. I will rebase and try this asap. Most likely ETA is this weekend.

**Comment by khronnuz** [comment]
@milessabin last build failed with:

```
java.lang.NumberFormatException: For input string: "2201313E.2201313E44"
	at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
	at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
...
```
https://travis-ci.org/milessabin/shapeless/jobs/133634220

Not sure what to do about it.

**Comment by InTheNow** [comment]
Looks OK now?

**Comment by milessabin** [comment]
Oh, cool! What (if anything) more needs to be done?

**Comment by InTheNow** [comment]
Does scoverage web actually get updated? I can't see anything, nor on this page - perhaps a missing GH hook?

**Comment by InTheNow** [comment]
https://codecov.io/gh/milessabin/shapeless/pull/551 :) Now just GH

**Comment by InTheNow** [comment]
Actually I think a bit more is needed - I'll comment in the files accordongly

File: .travis.yml [hunk 1]
```diff
@@ -14,7 +14,7 @@ cache:
     - $HOME/.sbt
 
 script:
-  - travis_wait 45 scripts/travis-publish.sh
+  - travis_wait 60 scripts/travis-publish.sh
```
Comment (InTheNow) [inline .travis.yml]: Add the following ([see](https://github.com/InTheNow/catalysts/blob/master/.travis.yml#L3-L4))

```yml
install:
  - pip install --user codecov
```

File: scripts/travis-publish.sh [hunk 2]
```diff
@@ -4,6 +4,7 @@
 # export TRAVIS_SCALA_VERSION=2.10.5;export TRAVIS_PULL_REQUEST="false";export TRAVIS_BRANCH="master"
 
 SBT="sbt ++${TRAVIS_SCALA_VERSION}"
+COVERAGE="$SBT clean coverage test coverageReport"
```
Comment (InTheNow) [inline scripts/travis-publish.sh]: [see](https://github.com/InTheNow/catalysts/blob/master/scripts/travis-publish.sh#L19)

`COVERAGE="$SBT clean coverage test coverageReport  && codecov"`

**Comment by khronnuz** [comment]
@InTheNow thanks for catching up those. I had this at some point (actually the curl `http... sh` method) but removed while testing the workarounds.

Last build passed. Do you prefer that I rebase again to squash into a single commit?

**Comment by InTheNow** [comment]
@khronnuz Ooh...it's working :+1: Yeah, I guess a squash would be neat. Nice job :smile:

**Comment by khronnuz** [comment]
@InTheNow done.

**Comment by InTheNow** [comment]
LGTM :+1: 

@milessabin ?

**Comment by milessabin** [comment]
Clicking through from the shield on the README I get to the report. It contains this message,
> Unable to retrieve commit from GitHub API. Please check that your bot user has authroization to milessabin/shapeless.

Any ideas?

**Comment by InTheNow** [comment]
Clicking on the README shield, I get the coverage report proper. Also, I see a nice coverage diff in https://github.com/milessabin/shapeless/pull/605.

Looks OK now?

**Comment by khronnuz** [comment]
Same for me. Worked both logged in and in an incognito window.

**Comment by milessabin** [comment]
Works for me now too ... must have been a transient thing.

Thanks both of you for sorting this out :-)

**Pull request merged**